upup/pkg/fi/cloudup/awstasks/iaminstanceprofile.go

```diff
@@ -93,17 +95,28 @@ func (e *IAMInstanceProfile) Run(c *fi.Context) error {
 	return fi.DefaultDeltaRunMethod(e, c)
 }
 
-func (s *IAMInstanceProfile) CheckChanges(a, e, changes *IAMInstanceProfile) error {
+func (_ *IAMInstanceProfile) ShouldCreate(a, e, changes *IAMInstanceProfile) (bool, error) {
+	if fi.BoolValue(e.Shared) {
```

Comment:
Instead of doing this, I think we should use the lifecycle options.

We still want to validate that the IAM role matches expectations, or else we have no way of verifying that it is correct.  So where we would set shared == true, we probably want Lifecycle to be LifecycleExistsAndValidates.

The tricky thing is it's probably going to make overriding it harder - I'm guessing we want some sort of oracle function for lifecycle cc @chrislovecnm

Reply:
@rifelpet take a look at https://github.com/kubernetes/kops/pull/4445
This will allow a user to set the lifecycle on whichever tasks they want to.  

@justinsb just thought of an edge case for this.  With this feature, you probably want to set the task`IAMRolePolicy` with an overridden lifecycle. I am thinking this may stop a user from having additional policies attached.  

I am thinking that https://github.com/rifelpet/kops/blob/12f72d405e06e47a74c82e60dd6b761253d60722/pkg/apis/kops/cluster.go#L160 will no longer work, which you override the lifecycle.

@rifelpet is what I am thinking, make any sense?

This edge case is fine since you can just define what you want for the IAMProfile, but we need to document this.  Not certain how my security group PR will act.

Reply:
@justinsb with IAM-related resources there are many ways they could be setup but still function as kops would expect. inline vs managed policies, a superset of expected policies, even inline policies named differently. How would the validation handle these scenarios? Are we sure its feasible to perform this validation effectively?

In my specific case, I will be using managed policies and attaching them to the roles. Would the validation step play nicely with that?

@chrislovecnm does #3453 implement the lifecycle changes based on #4445 ? If so I can mimic the similar changes here, otherwise I'm not sure exactly what needs to happen in the code to accomplish what you're describing. (Not too familiar with the relationship between lifecycles and tasks, but ill take a closer look hopefully later today)

Reply:
Oh, unless you're implying that validation simply checks for the existence of the instance profile. That makes more sense. Do we have other tasks that perform validation on LifecycleExistsAndValidates that I could use as an example? I'm not seeing any, but basically we'll just want to make sure `findIAMInstanceProfile` doesnt return nil. Is that correct?